File: lib/utils/com_scanner.dart
import 'package:flutter_libserialport/flutter_libserialport.dart';

class DongleDeviceInfo {
  final String portName;
  final String? productName;
  final int? vendorId;
  final int? productId;

  DongleDeviceInfo({
    required this.portName,
    this.productName,
    this.vendorId,
    this.productId,
  });
}

class ComScanner {
  // 🎯 這是 Silicon Labs CP210x 常見的 VID
  static const int _siliconLabsVid = 0x10C4;

  /// 掃描並過濾出所有的 Silicon Labs Dongles
  static List<DongleDeviceInfo> findDonglePorts() {
    List<DongleDeviceInfo> foundDongles = [];

    // 1. 取得系統所有可用序列埠的 ID 名稱 (例如 COM3, COM4)
    final List<String> availablePorts = SerialPort.availablePorts;

    for (final name in availablePorts) {
      final port = SerialPort(name);

      try {
        // 2. 取得硬體細節 (對應 Go 版的 GetDetailedPortsList)
        final int? vid = port.vendorId;
        final int? pid = port.productId;
        final String? product = port.productName;

        // 3. 過濾邏輯 (對應 Go 版的 strings.Contains)
        bool isDongle = false;

        // 條件 A: 檢查 VID 是否為 Silicon Labs (10C4)
        if (vid == _siliconLabsVid) {
          isDongle = true;
        }
        // 條件 B: 檢查產品名稱關鍵字
        else if (product != null) {
          final pUpper = product.toUpperCase();
          if (pUpper.contains("SILICON LABS") || pUpper.contains("CP210X")) {
            isDongle = true;
          }
        }

        if (isDongle) {
          foundDongles.add(
            DongleDeviceInfo(
              portName: name,
              productName: product,
              vendorId: vid,
              productId: pid,
            ),
          );
        }
      } catch (e) {
        // 有些 Port 可能被其他程式佔用無法讀取描述，直接跳過
        continue;
      } finally {
        // 釋放資源，不要開著 Port，只做掃描
        port.dispose();
      }
    }

    return foundDongles;
  }
}

File: go_core/com_scanner.go
package main

import (
	"fmt"
	"strings"

	"go.bug.st/serial/enumerator"
)

// FindDonglePorts 掃描系統中所有的 COM Port，並過濾出 Silicon Labs 的裝置
func FindDonglePorts() []string {
	fmt.Println("🔍 正在掃描系統 COM Port...")

	// 取得詳細的 Port 列表
	ports, err := enumerator.GetDetailedPortsList()
	if err != nil {
		fmt.Printf("❌ 掃描失敗: %v\n", err)
		return []string{}
	}

	var foundPorts []string

	if len(ports) == 0 {
		fmt.Println("⚠️  未偵測到任何 COM Port")
		return foundPorts
	}

	for _, port := range ports {
		// 印出所有找到的 Port 資訊 (除錯用，之後可以註解掉)
		fmt.Printf("   Found: %s | Product: %s | VID/PID: %s\n", port.Name, port.Product, port.VID+"/"+port.PID)

		// 過濾條件：你的截圖顯示名稱包含 "Silicon Labs" 或 "CP210x"
		// 我們把它轉成大寫來比對，比較保險
		productName := strings.ToUpper(port.Product)

		// 這裡設定關鍵字，只要名稱包含這些就會被選中
		if strings.Contains(productName, "SILICON LABS") || strings.Contains(productName, "CP210X") {
			foundPorts = append(foundPorts, port.Name)
			fmt.Printf("   ✅ 識別到 Dongle: %s (%s)\n", port.Name, port.Product)
		}
	}

	fmt.Printf("📊 掃描完成，共找到 %d 個有效 Dongle\n", len(foundPorts))
	return foundPorts
}
